### PR TITLE
[Clang] Introduce 'clang-nvlink-wrapper' to work around 'nvlink'

### PR DESCRIPTION
**Pull request by jhuber6**
Summary:
The `clang-nvlink-wrapper` is a utility that I removed awhile back
during the transition to the new driver. This patch adds back in a new,
upgraded version that does LTO + archive linking. It's not an easy
choice to reintroduce something I happily deleted, but this is the only
way to move forward with improving GPU support in LLVM.

While NVIDIA provides a linker called 'nvlink', its main interface is
very difficult to work with. It does not provide LTO, or static linking,
requires all files to be named a non-standard `.cubin`, and rejects link
jobs that other linkers would be fine with (i.e empty). I have spent a
great deal of time hacking around this in the GPU `libc` implementation,
where I deliberately avoid LTO and static linking and have about 100
lines of hacky CMake dedicated to storing these files in a format that
the clang-linker-wrapper accepts to avoid this limitation.

The main reason I want to re-intorudce this tool is because I am
planning on creating a more standard C/C++ toolchain for GPUs to use.
This will install files like the following.
```
<install>/lib/nvptx64-nvidia-cuda/libc.a
<install>/lib/nvptx64-nvidia-cuda/libc++.a
<install>/lib/nvptx64-nvidia-cuda/libomp.a
<install>/lib/clang/19/lib/nvptx64-nvidia-cuda/libclang_rt.builtins.a
```
Linking in these libraries will then simply require passing `-lc` like
is already done for non-GPU toolchains. However, this doesn't work with
the currently deficient `nvlink` linker, so I consider this a blocking
issue to massively improving the state of building GPU libraries.

In the future we may be able to convince NVIDIA to port their linker to
`ld.lld`, but for now this is the only workable solution that allows us
to hack around the weird behavior of their closed-source software.
This also copies some amount of logic from the clang-linker-wrapper,
but not enough for it to be worthwhile to merge them I feel. In the
future it may be possible to delete that handling from there entirely.

### CODE REVIEW

**Comment by jlebar** [comment]
@Artem-B asked me to review nvptx patches while he's OOO, but this one is pretty far outside my depth.  Are you OK waiting until he's back?  I don't know exactly when that will be, but based on his IMs to me, he should be back early July.

**Comment by jhuber6** [comment]
> @Artem-B asked me to review nvptx patches while he's OOO, but this one is pretty far outside my depth. Are you OK waiting until he's back? I don't know exactly when that will be, but based on his IMs to me, he should be back early July.

No problem, I knew that it would probably take awhile to get reviewed given the size. I believe he said he'd be back early July as well, so maybe next week? It'd probably require his input, along with some of the other interested parties in clang to see how they feel about reviving one of these old tools.

(However if you know anything about the NVPTX varargs API I think https://github.com/llvm/llvm-project/pull/96015 is mostly just waiting for someone to say that it's a mostly correct lowering)

**Comment by jhuber6** [comment]
@MaskRay So, I think my symbol resolution is (unsurprisingly) subtly broken. Is there a canonical way to handle this? I first thought that we could simply perform the symbol resolutions as normal for every file, but keep track of which symbols were "lazy". However, I couldn't figure out how to then tell if a lazy symbol should be extracted or not because there's no information on which files use which symbols. Maybe I just scan all the files and see if they reference a symbol that's marked defined and lazy?

**Comment by jhuber6** [comment]
Re-did it and tested it against `libc` in https://github.com/llvm/llvm-project/pull/96972 so it will have a CI running it one  that lands. it works for other cases I've tested, but let me know if something else should be added.

**Comment by jhuber6** [comment]
Ping

**Comment by jlebar** [comment]
@Artem-B

File: clang/docs/ClangNVLinkWrapper.rst
Comment (Artem-B) [inline clang/docs/ClangNVLinkWrapper.rst:17]: This is written from a standpoint of someone already familiar with nvlink quirks. For the rest of the world "support fo static linking with .a and object files ending in .o" will be somewhat baffling. "Isn't how things already work?"

I'd rephrase it a bit along the lines of. "nvlink has a number of known quirks that make its use in a unified HIP/CUDA/OpenMP compilation pipeline problematic" followed by the list of those quirks. The wrapper isolates those differences and presents an interface close enough to ld.lld to simplify the integration.

Does it make sense?

File: clang/docs/ClangNVLinkWrapper.rst
Comment (Artem-B) [inline clang/docs/ClangNVLinkWrapper.rst:60]: Does it mean "intended to be invoked **only** from direct compilation targeting NVPTX" ? Or do you foresee other potential use cases?

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:236]: The comment does not mention `.lib`. Where do those `.lib`  inputs come from?

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:250]: I think we could just use `llvm::errs() << llvm::join(std::next(CmdArgs.begin()), CmdArgs.end(), " ") << "\n"`

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:313]: Silently ignoring `-O*` when -g is specified is probably not the best way to handle thing here.

I would prefer to downgrade `-g` to `-lineinfo` if optimization options other than `-O0` are specified. Reduced debug granularity is likely closer to what user intended. Nobody expects good debug info if optimization is enabled, and if debug build is done with optimization, presumably the optimization is important.

A warning may be useful, too.

**Review by Artem-B — COMMENTED** [review]
First batch of comments on the patch -- I only got till about the middle of ClangNVLinkWrapper.cpp. Will continue reviewing tomorrow.

File: clang/docs/ClangNVLinkWrapper.rst
Comment (jhuber6) [reply clang/docs/ClangNVLinkWrapper.rst:17]: Yep, that's fair. I'll try to rewrite it to be more clear.

File: clang/docs/ClangNVLinkWrapper.rst
Comment (jhuber6) [reply clang/docs/ClangNVLinkWrapper.rst:60]: I mean this as a stand-in for `--target=nvptx64-nvidia-cuda`, which is what direct compilation uses, as does the `clang-linker-wrapper` to generate the link job. So, it's pretty much just a way to say "This is for times when the NVPTX toolchain needs a functional linker".

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:250]: Good point.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:236]: I should remove that, it's for Windows compatibility with their search paths but I don't think we need it here.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:313]: We can't downgrade once it gets here, we just need to assume that if there's debugging using any optimization level will make `ptxas` die. I can definitely see a warning though.

**Comment by jhuber6** [comment]
> First batch of comments on the patch -- I only got till about the middle of ClangNVLinkWrapper.cpp. Will continue reviewing tomorrow.

Appreciate it. Sorry about dropping a huge patch like this on you.

**Comment by jhuber6** [comment]
The CI seems to fail because it's not building the tool yet, the patch should enable that dependency but maybe the bot doesn't know how to pick it up until it lands.

**Comment by jhuber6** [comment]
ping

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:296]: We also have `--ptxas-path=` that we may want to propagate here from the top-level compilation:

https://github.com/llvm/llvm-project/blob/25897ba420b93449542a60ffc9542e62b7480807/clang/include/clang/Driver/Options.td#L1340

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:356]: If we do expect an integer value, we should use a standard function for conversion.

Also, what if we get `-Os` ?

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:368]: Does it handle 'magic' file names like `-` ? Or do we want or need to handle it manually?

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:659]: I suspect this will be potentially problematic for large distributed builds, where the worker job doing the linking may not have enough scratch space to fit both the output and a copy of all inputs. We may want to eventually add platform-specific optimizations here. E.g, maybe create a `.cubin` symlinks to the actual inputs on UNIX-like hosts?

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:751]: Should we exit here? I think that's the 'standard' expectation. At least clang does so.

**Review by Artem-B — APPROVED** [review]
LGTM for the patch in general, though I can't vouch for the details of the linking process. I'll defer to @MaskRay on that.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:356]: `-Os` is converted to `-O2` by clang for LTO purposes

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:296]: Good point, didn't know those were separate.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:368]: I think LLVM handles `-` but I've never tried it (Though I can't imagine many users want to output binary to stdout).

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:659]: Yeah, I was concerned with that as well. I think a symlink would work well for this since only the LTO files need to truly be copied.  I can add that later.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:751]: So, I think I made it not exit just so that it would also print `nvlink`'s version.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (Artem-B) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:356]: Fair enough, In that case, if -Os is unknown/unsupported, it should probably be rejected.

**Comment by dyung** [comment]
@jhuber6 this change seems to be causing build failures on some bots:
- https://lab.llvm.org/buildbot/#/builders/174/builds/2129
- https://lab.llvm.org/buildbot/#/builders/190/builds/2383

Can you take a look?

**Comment by dyung** [comment]
> @jhuber6 this change seems to be causing build failures on some bots:
> 
> * https://lab.llvm.org/buildbot/#/builders/174/builds/2129
> * https://lab.llvm.org/buildbot/#/builders/190/builds/2383
> 
> Can you take a look?

Nevermind, I just noticed you already fixed it. Sorry for the noise!

**Comment by jhuber6** [comment]
> @jhuber6 this change seems to be causing build failures on some bots:
> 
>     * https://lab.llvm.org/buildbot/#/builders/174/builds/2129
> 
>     * https://lab.llvm.org/buildbot/#/builders/190/builds/2383
> 
> 
> Can you take a look?

Gah, sorry forgot to suppress that error when it's just running tests. I'll have it fixed in a few minutes or revert it if I can't fix it fast enough.

**Comment by dyung** [comment]
https://lab.llvm.org/buildbot/#/builders/174/builds/2130

Actually I noticed (and I think you already did as well), that there seems to be a test failure of nvwrapper.c after you fixed the build error. Just wanted to let you know in case you were not already aware.

**Comment by jhuber6** [comment]
> https://lab.llvm.org/buildbot/#/builders/174/builds/2130
> 
> Actually I noticed (and I think you already did as well), that there seems to be a test failure of nvwrapper.c after you fixed the build error. Just wanted to let you know in case you were not already aware.

Yep, it's caused by the program lookup not finding `nvlink` because I sometimes forget that not everything has CUDA installed.

**Comment by MaskRay** [comment]
clang/test/CMakeLists.txt CLANG_TEST_DEPS also needs an update. `

The test might fail due to `/tmp/Rel/bin/clang-nvlink-wrapper: error: Unable to find 'nvlink' in path`

**Comment by jhuber6** [comment]
> clang/test/CMakeLists.txt CLANG_TEST_DEPS also needs an update. `
> 
> The test might fail due to `/tmp/Rel/bin/clang-nvlink-wrapper: error: Unable to find 'nvlink' in path`

Also noticed that one, just pushed a fix a minute ago. Sorry for the mess.

**Comment by dyung** [comment]
@jhuber6 it looks like there still might be build failures on Windows:
https://lab.llvm.org/buildbot/#/builders/46/builds/2077
```
FAILED: tools/clang/tools/clang-nvlink-wrapper/CMakeFiles/clang-nvlink-wrapper.dir/ClangNVLinkWrapper.cpp.obj 
C:\bin\ccache.exe C:\PROGRA~2\MICROS~1\2019\BUILDT~1\VC\Tools\MSVC\1429~1.301\bin\HostX64\x64\cl.exe  /nologo /TP -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Itools\clang\tools\clang-nvlink-wrapper -IZ:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\tools\clang-nvlink-wrapper -IZ:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\include -Itools\clang\include -Iinclude -IZ:\b\llvm-clang-x86_64-sie-win\llvm-project\llvm\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -g -O0 -std:c++17 /showIncludes /Fotools\clang\tools\clang-nvlink-wrapper\CMakeFiles\clang-nvlink-wrapper.dir\ClangNVLinkWrapper.cpp.obj /Fdtools\clang\tools\clang-nvlink-wrapper\CMakeFiles\clang-nvlink-wrapper.dir\ /FS -c Z:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\tools\clang-nvlink-wrapper\ClangNVLinkWrapper.cpp
cl : Command line warning D9002 : ignoring unknown option '-g'
cl : Command line warning D9002 : ignoring unknown option '-O0'
Z:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\tools\clang-nvlink-wrapper\ClangNVLinkWrapper.cpp(429): error C2440: 'initializing': cannot convert from 'const ValueTy' to '_Ty2 &&'
        with
        [
            ValueTy=`anonymous-namespace'::Symbol
        ]
        and
        [
            _Ty2=`anonymous-namespace'::Symbol
        ]
Z:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\tools\clang-nvlink-wrapper\ClangNVLinkWrapper.cpp(429): note: Conversion loses qualifiers
Z:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\tools\clang-nvlink-wrapper\ClangNVLinkWrapper.cpp(459): error C2440: 'initializing': cannot convert from 'const ValueTy' to '_Ty2 &&'
        with
        [
            ValueTy=`anonymous-namespace'::Symbol
        ]
        and
        [
            _Ty2=`anonymous-namespace'::Symbol
        ]
Z:\b\llvm-clang-x86_64-sie-win\llvm-project\clang\tools\clang-nvlink-wrapper\ClangNVLinkWrapper.cpp(459): note: Conversion loses qualifiers
```

File: clang/tools/clang-nvlink-wrapper/CMakeLists.txt
Comment (bogner) [inline clang/tools/clang-nvlink-wrapper/CMakeLists.txt:39]: This looks suspicious. Do we actually want to build this with `-g -O0` all the time or was this left in from debugging or something like that? In the unlikely event that we do want this for some reason, it won't work as is on windows anyway.

Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/CMakeLists.txt:39]: Whoops, thanks for pointing that out.

**Comment by nico** [comment]
Why do we need a new binary for this, instead of having something like `clang -cc1_nvlink` that calls a custom mode within clang?

And if there's a good reason for that, could clang-linker-wrapper and clang-nvlink-wrapper at least be the same binary?

**Comment by vvereschaka** [comment]
@jhuber6 ,
there is still few failed bot because of these changes
* https://lab.llvm.org/buildbot/#/builders/193/builds/1224
* https://lab.llvm.org/buildbot/#/builders/2/builds/2790

would you fix the problem or revert the changes?

**Comment by jhuber6** [comment]
> @jhuber6 , there is still few failed bot because of these changes
> 
>     * https://lab.llvm.org/buildbot/#/builders/193/builds/1224
> 
>     * https://lab.llvm.org/buildbot/#/builders/2/builds/2790
> 
> 
> would you fix the problem or revert the changes?

Looks like it's complaining about the const-ness of the value decomposition? It's fine one my end so I don't know if it's a Windows thing, I'll just try putting `const` there and seeing if that fixes it.

**Comment by vvereschaka** [comment]
Thank you @jhuber6 
both of those failed builders got fixed.

**Comment by Artem-B** [comment]
@nico: 
> Why do we need a new binary for this, instead of having something like `clang -cc1_nvlink` that calls a custom mode within clang?

Do we have existing precedents for such built-in tools, other than cc1 itself? If the linker wrapper can be part of clang itself, it would make some things easier logistically.

File: clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp
Comment (MaskRay) [inline clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:54]: For new user-facing tools, ideally use OptTable instead of cl::opt. The latter has lots of minor issues, e.g. not differentiating single/double-dash options.

Comment (jhuber6) [reply clang/tools/clang-nvlink-wrapper/ClangNVLinkWrapper.cpp:54]: So, most everything uses the OptTable interface, but there's a few options like these which each tool individually seems to define that interface w/ LLVM. It's pretty much required to give the linker similar semantics to `opt` when you pass `-mllvm` to it I believe.

**Comment by jhuber6** [comment]
> Why do we need a new binary for this, instead of having something like `clang -cc1_nvlink` that calls a custom mode within clang?
> 
> And if there's a good reason for that, could clang-linker-wrapper and clang-nvlink-wrapper at least be the same binary?

Sorry, missed this in the deluge of build failures. I don't think this really fits with a different `-cc1` mode because it's just a linker step that we augment. To me, this is like suggesting we merge `ld.lld` into `clang` so we can do everything in a single invocation. It's certainly possible, but I think this fits alright into the workflow.

**Pull request merged**
